Unicode/logger fix + bugfix: salt/utils/parsers

isbm opened the pull request:
### What does this PR do?

Bugfix/refactoring

### What issues does this PR fix or reference?

- Fixes logging with unicode
- Fixes PID file deletion as non-root (Master)
- Fixes unit tests for this topic

### Tests written?

Yes

Commit:
Add logging

Commit:
Suppress logging PID file removal if running on non-root.

Commit:
Lintfix: remove extra-parenthesis. Remove unnecessary back-slashes

Commit:
Rename logger variable

Commit:
Remove extra-parenthesis and extra-backslashes

Commit:
Remove extra-backslashes

Commit:
Use logger variable

Commit:
Fix unicode support for the logging

Commit:
Use logger variable and fix unicode support

Commit:
Bugfix: do not crash on import error, but log that incident instead

Commit:
PEP8: remove extra-parenthesis

Commit:
Fix logging, remove extra-parenthesis (PEP8)

Commit:
Use pytest

Commit:
Add tear-down method

Commit:
Fix unit test to attempt remove PID file as root

Commit:
Add unit test for calling PID deletion as non-root user

Commit:
Add debugging information to the log

Commit:
Remove unnecessary string on get->return

Commit:
Throw away unnecessary variable handling

Commit:
Add debug logging call to the unit test check

Commit:
Refactor a unit test for calling PID file removal without the exception

Commit:
Remove unused variables, rename mixin variable to more exact name

isbm commented:
@rallytime @terminalmage please take a look.

Commit:
Fix lints

terminalmage commented on salt/utils/parsers.py:
```diff
@@ -1044,12 +1037,13 @@ def start(self):
+        msg = [six.text_type(self.__class__.__name__)]
         if signum == signal.SIGINT:
-            msg += ' received a SIGINT.'
+            msg.append('received a SIGINT.')
```
The string types do not have an `append` function.

terminalmage replied:
I missed the brackets around msg when it was being initialized. Still, allocating a list for this is slower:

```python
>>> timeit.Timer('''
... x = ['foo']
... x.append('bar')
... x = ' '.join(x)''').repeat()
[0.22678208351135254, 0.21661591529846191, 0.20710206031799316]
>>> timeit.Timer('''
... x = 'foo'
... x += " bar"''').repeat()
[0.08382701873779297, 0.054234981536865234, 0.05179405212402344]
```

terminalmage commented on salt/utils/parsers.py:
```diff
@@ -1044,12 +1037,13 @@ def start(self):
         elif signum == signal.SIGTERM:
-            msg += ' received a SIGTERM.'
-        logging.getLogger(__name__).warning('%s Exiting.', msg)
+            msg.append('received a SIGTERM.')
```
Same thing here, this line needs to be reverted.

isbm replied:
It is earlier an array... :wink:

terminalmage commented on salt/utils/parsers.py:
```diff
@@ -2739,31 +2733,31 @@ def _find_raet_minion(self, opts):
+                if os.path.exists(ha) and not os.path.isfile(ha) and not os.path.isdir(ha):  # minion manor yard
+                    return True
+            except ImportError as ex:
+                logger.error("Error while importing Yard: %s", six.text_type(ex))
```
No need for `six.text_type(ex)`, this can just be `ex`. String replacement will automagically run `ex.__unicode__()` to get a string, because the string with the replacement placeholders is a unicode. For example:

```python
>>> u'The error is: %s' % TypeError(u'питон')
u'The error is: \u043f\u0438\u0442\u043e\u043d'
```

Commit:
Drop forcing to unicode exception object

rallytime commented:
@isbm Looks like we have a related test failure here: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu16-py3/6776/

Can you take a look?

isbm commented:
@rallytime Seems like your test suite on Ubuntu 16 is clashing with the Mock 2.0.0 from Pip, saying in logs it is installed, but actually using just a standard mock library from Python 3.5 where this method is actually absent. This method is [new to Python 3.6](https://docs.python.org/release/3.6.0/library/unittest.mock.html#unittest.mock.Mock.assert_called).

Seems we found a bug in test setup!

Commit:
Use mock 2.0.0 instead of unittest.mock on Python 3 (#2)

This preserves the custom mock_open we backported from unittest.mock,
but otherwise ditches unittest.mock as it does not have
MagicMock.assert_called in Python releases before 3.6.

This allows us to maintain a uniform mock version across all platforms
and Python releases.

terminalmage commented:
Via https://github.com/saltstack/salt-jenkins/pull/840, https://github.com/saltstack/salt-jenkins/pull/841, and the commit just merged into this PR, we should now be using mock 2.0.0 from PyPI for everything but `mock_open`, which we have backported.

isbm commented:
@terminalmage @rallytime seems working now. I am wondering why Cent7/U14 still fails, but that's seems unrelated to here.

Thanks!

rallytime commented:
@isbm Ah, yeah the Ubuntu 14 errors are unrelated. But the cp test failures on the PY3 build are related to the concerns @gtmanfred brought up in https://github.com/saltstack/salt-jenkins/pull/840

@terminalmage Can you take another look into the changes you made in our custom `mock_open`?

gtmanfred commented:
Actually, those cp changes are unrelated.  mock_open was affecting the cp.push test

I have been seeing this same error with repo.saltstack.com, the 599 error, since yesterday when we switched the repo to be behind a DO load balancer instead of haproxy.

terminalmage commented:
I didn't change our backport, I just changed the bits that import directly from unittest.mock on Python 3. I'll look at the test failures.

rallytime commented:
Ah, OK. Thanks for the explanation everyone.
